File: GameServer/BattleSimulation/BattleAttributes.cs
using GameServer.Models.Attributes;
using static GameServer.AttributeType;

namespace GameServer.BattleSimulation
{
    public class BattleAttributes
    {
        private static readonly int _attributesMaxId = (int)Enum.GetValues(typeof(AttributeType)).GetValue(Enum.GetValues(typeof(AttributeType)).Length - 1);
        private readonly List<decimal> _attributes;

        //public double this[int index]
        //{
        //    get => (double)_attributes[index];
        //    set => _attributes[index] = (decimal)value;
        //}

        public double this[AttributeType index]
        {
            get => (double)_attributes[(int)index];
            set => _attributes[(int)index] = (decimal)value;
        }

        public BattleAttributes(IEnumerable<BattlerAttribute> atts)
        {
            _attributes = GetEmptyAttributeList();
            foreach (var att in atts)
            {
                _attributes[(int)att.AttributeId] += att.Amount;
            }
            CalculateDerivedValues();
        }

        private static List<decimal> GetEmptyAttributeList()
        {
            return Enumerable.Repeat(0m, _attributesMaxId).ToList();
        }

        private void CalculateDerivedValues()
        {
            this[MaxHealth] += 50.0 + (20.0 * this[Endurance]) + (5 * this[Strength]);
            this[Defense] += 2.0 + this[Endurance] + (0.5 * this[Agility]);
            this[CooldownRecovery] += (0.4 * this[Agility]) + (0.1 * this[Dexterity]);
            this[DropBonus] += Math.Log10(this[Luck]);
            //this[CriticalChance] = 0.0;
            //this[CriticalDamage] = 0.0;
            //this[DodgeChance] = 0.0;
            //this[BlockChance] = 0.0;
            //this[BlockReduction] = 0.0;
        }
    }
}

File: GameServer/TypeScript/Game/Battle/BattleAttributes.ts
import { EAttribute } from "../Shared/Api/Enums";
import { IBattlerAttribute } from "../Shared/Api/Types";
import { normalizeText } from "../Shared/GlobalFunctions";

export class BattleAttributes {
    static #attributesMaxId = Object.values(EAttribute)[Object.values(EAttribute).length - 1] as number;
    attributes: number[];
    //battlerAttributes: IBattlerAttribute[];

    constructor(attList: IBattlerAttribute[], calcDerivedStats: boolean = true) {
        //this.battlerAttributes = attList;
        this.attributes = new Array<number>(BattleAttributes.#attributesMaxId);
        this.attributes.fill(0, 0, BattleAttributes.#attributesMaxId);
        attList.forEach(att => this.attributes[att.attributeId] += att.amount);
        if (calcDerivedStats) {
            this.#calculateDerivedStats()
        }
    }

    getValue(attId: number) {
        return this.attributes[attId];
    }

    // resetAttributes(calcDerivedStats: boolean = true) {
    //     this.attributes.fill(0, 0, BattleAttributes.#attributesMaxId);
    //     this.battlerAttributes.forEach(att => this.attributes[att.attributeId] += att.amount);
    //     if (calcDerivedStats) {
    //         this.#calculateDerivedStats()
    //     }
    // }

    // addAttributes(attList: IBattlerAttribute[], calcDerivedStats: boolean = true) {
    //     this.battlerAttributes.push(...attList);
    //     this.resetAttributes(calcDerivedStats);
    // }

    getAttributeMap(includeZeroes: boolean = false) {
        return this.attributes
            .map((att, i) => ({name: normalizeText(EAttribute[i]), value: att}))
            .filter(att => att.value != 0 || includeZeroes);
    }

    #calculateDerivedStats() {
        this.attributes[EAttribute.MaxHealth] += 50 + (20 * this.attributes[EAttribute.Endurance]) + (5 * this.attributes[EAttribute.Strength]);
        this.attributes[EAttribute.Defense] += 2 + this.attributes[EAttribute.Endurance] + (0.5 * this.attributes[EAttribute.Agility]);
        this.attributes[EAttribute.CooldownRecovery] += (0.4 * this.attributes[EAttribute.Agility]) + (0.1 * this.attributes[EAttribute.Dexterity]);
        this.attributes[EAttribute.DropBonus] += Math.log10(this.attributes[EAttribute.Luck]);
        // this.attributes[EAttribute.CriticalChance] = 0;
        // this.attributes[EAttribute.CriticalDamage] = 0;
        // this.attributes[EAttribute.DodgeChance] = 0;
        // this.attributes[EAttribute.BlockChance] = 0;
        // this.attributes[EAttribute.BlockReduction] = 0;
    }
}